pilot/pkg/trustbundle/trustbundle.go

```diff
@@ -0,0 +1,181 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package trustbundle
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"strings"
+	"sync"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+	"istio.io/pkg/log"
+)
+
+// source is all possible sources of MeshConfig
+type Source int
+
+type TrustAnchorConfig struct {
+	Source Source
+	Certs  []string
+}
+
+type TrustAnchorUpdate struct {
+	TrustAnchorConfig
+}
+
+type TrustBundle struct {
+	sourceConfig map[Source]TrustAnchorConfig
+	mutex        sync.RWMutex
+	mergedCerts  []string
+	updatecb     func()
+}
+
+var (
+	trustBundleLog = log.RegisterScope("trustBundle", "Workload mTLS trust bundle logs", 0)
+
+	// GlobalWorkloadTrustbundle Needed to interface with XDS
+	GlobalWorkloadTrustbundle *TrustBundle = nil
+)
+
+const (
+	SourceIstioCA Source = iota
+	SourceMeshConfig
+	SourceIstioRA
+	// SOURCE_SPIFFE_ENDPOINT
+	UpdateChannelLen = 10
+)
+
+func checkSameCerts(certs1 []string, certs2 []string) bool {
+	if len(certs1) != len(certs2) {
+		return false
+	}
+	for i := range certs1 {
+		if certs1[i] != certs2[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func NewTrustBundle(updatecb func()) *TrustBundle {
+	tb := &TrustBundle{
+		sourceConfig: map[Source]TrustAnchorConfig{
+			SourceIstioCA:    {Source: SourceIstioCA, Certs: []string{}},
```

Comment:
nit: do we need `Source` in TrustAnchorConfig given its already in the map? Just feels a bit repetative. Could we have a map of `Source->Certs` instead?